src/reflect/scala/reflect/internal/tpe/GlbLubs.scala

```diff
@@ -482,20 +482,26 @@ private[internal] trait GlbLubs {
       try {
         val (ts, tparams) = stripExistentialsAndTypeVars(ts0)
         val glbOwner = commonOwner(ts)
-        def refinedToParents(t: Type): List[Type] = t match {
-          case RefinedType(ps, _) => ps flatMap refinedToParents
-          case _ => List(t)
+        def refinedToParentsList(ts: List[Type]): List[Type] = {
+          val res = mutable.ListBuffer.empty[Type]
+          def loop(ty: Type): Unit = ty match {
+            case RefinedType(ps, _) => ps.foreach(loop)
+            case _ => res += ty
+          }
+          ts foreach loop
+          res.toList
         }
-        def refinedToDecls(t: Type): List[Scope] = t match {
+        def refinedDeclsForeach(ty: Type)(cont: Scope => Unit): Unit = ty match {
           case RefinedType(ps, decls) =>
-            val dss = ps flatMap refinedToDecls
-            if (decls.isEmpty) dss else decls :: dss
-          case _ => List()
+            ps foreach (t => refinedDeclsForeach(t)(cont))
+            if (! decls.isEmpty)
+              cont(decls)
+          case _ =>
         }
-        val ts1 = ts flatMap refinedToParents
-        val glbBase = intersectionType(ts1, glbOwner)
+        val ts1 = refinedToParentsList(ts)
```

Comment:
```
        val ts1 = {
          def refinedToParentsList(ts: List[Type]): List[Type] = {
            val res = mutable.ListBuffer.empty[Type]
            def loop(ty: Type): Unit = ty match {
              case RefinedType(ps, _) => ps.foreach(loop)
              case _ => res += ty
            }
            ts foreach loop
            res.toList
          }
          refinedToParentsList(ts)
        }
```

Reply:
Under the criteria to bring it closer to call site, these two functions could be moved closer to the call sites. On the other hand, they could also be moved further away, to a wider scope, so as to indicate that their code does not depend on any local variable of the `glb` function: it is just providing a way to, recursively, run through the parents of a list.